SourceX/controls/plrctrls.cpp

```diff
@@ -0,0 +1,693 @@
+		if (newCurHidden) { // show cursor first, before clicking
+			SetCursor_(CURSOR_HAND);
+			newCurHidden = false;
+		} else if (spselflag) {
+			SetSpell();
+		} else {
+			LeftMouseCmd(false);
+		}
+	}
+}
+
+// Move map or mouse no more than 60 times per second.
+// This is called from both the game loop and the event loop for responsiveness.
+void HandleRightStickMotionAt60Fps()
+{
+	static std::int64_t currentTime = 0;
+	static std::int64_t lastTime = 0;
+	currentTime = SDL_GetTicks();
+
+	if (currentTime - lastTime > 15) {
+		HandleRightStickMotion();
+		lastTime = currentTime;
+	}
+}
+
+struct RightStickAccumulator {
+	void start(int *x, int *y)
+	{
+		hiresDX += rightStickX * kGranularity;
+		hiresDY += rightStickY * kGranularity;
+		*x += hiresDX / slowdown;
+		*y += -hiresDY / slowdown;
+	}
+
+	void finish()
+	{
+		// keep track of remainder for sub-pixel motion
+		hiresDX %= slowdown;
+		hiresDY %= slowdown;
+	}
+
+	static const int kGranularity = (1 << 15) - 1;
+	int slowdown; // < kGranularity
+	int hiresDX;
+	int hiresDY;
+};
+
+} // namespace
+
+void HandleRightStickMotion()
+{
+	// deadzone is handled in ScaleJoystickAxes() already
+	if (rightStickX == 0 && rightStickY == 0)
+		return;
+
+	if (automapflag) { // move map
+		static RightStickAccumulator acc = { /*slowdown=*/(1 << 14) + (1 << 13), 0, 0 };
+		int dx = 0, dy = 0;
+		acc.start(&dx, &dy);
+		if (dy > 1)
+			AutomapUp();
+		else if (dy < -1)
+			AutomapDown();
+		else if (dx < -1)
+			AutomapRight();
+		else if (dx > 1)
+			AutomapLeft();
+		acc.finish();
+	} else { // move cursor
+		if (pcurs == CURSOR_NONE) {
+			SetCursor_(CURSOR_HAND);
+			newCurHidden = false;
+		}
+		static RightStickAccumulator acc = { /*slowdown=*/(1 << 13) + (1 << 12), 0, 0 };
+		int x = MouseX;
+		int y = MouseY;
+		acc.start(&x, &y);
+		if (x < 0)
+			x = 0;
+		if (y < 0)
+			y = 0;
+		SetCursorPos(x, y);
+		acc.finish();
+	}
+}
+
+void plrctrls_after_check_curs_move()
+{
+	HandleRightStickMotionAt60Fps();
+
+	// check for monsters first, then items, then towners.
+	if (pcurs <= 0) { // cursor should be missing
+		if (!checkMonstersNearby(false)) {
+			pcursmonst = -1;
+			checkItemsNearby(false);
+			checkTownersNearby(false);
+		} else {
+			pcursitem = -1;
+		}
+	}
+}
+
+void plrctrls_after_game_logic()
+{
+	movement();
+}
+
+void useBeltPotion(bool mana)
+{
+	static DWORD menuopenslow = 0;
+	DWORD ticks = GetTickCount();
+	int invNum = 0;
+	if (ticks - menuopenslow < 300) {
+		return;
+	}
+	menuopenslow = ticks;
+	for (int i = 0; i < MAXBELTITEMS; i++) {
+		const auto id = AllItemsList[plr[myplr].SpdList[i].IDidx].iMiscId;
+		if ((!mana && (id == IMISC_HEAL || id == IMISC_FULLHEAL)) || (mana && (id == IMISC_MANA || id == IMISC_FULLMANA)) || id == IMISC_REJUV || id == IMISC_FULLREJUV) {
+			if (plr[myplr].SpdList[i]._itype > -1) {
+				invNum = i + INVITEM_BELT_FIRST;
+				UseInvItem(myplr, invNum);
+				break;
+			}
+		}
+	}
+}
+
+void performPrimaryAction()

```

Comment:
The heavy nesting in this function makes it hard to read and also the delays seam arbitrery 300, 400, 500, 600?